### ISSUE TITLE
Debounce: only trigger reload at end of last call

### ISSUE THREAD

**Issue by adambuczynski** [BrowserSync/browser-sync#982]
Currently, it seems that browser sync *immediately* issues a reload on the first file change event, even with the `reloadDebounce` option set to 1000ms. This helps prevent subsequent reloads when more files change, but it reloads the browser too soon, because ideally it should only refresh once all file changes are done. E.g. the reload call should only be issues when all debounced events are finished.

A combination of `reloadDebounce` and `reloadDelay` also doesn't yield the desired result. It merely delays the first reload call by the specified delay, but if file changes happen after that delay they are not included in the reload.

As such, it doesn't seem very useful for an integrated build process. Is there another setting for this I've missed?

Ideally, I'd want the following to happen:

1. First file is changed, BS waits by the debounce time to see if another file change request is registered.
2. If another file is changed, BS resets the debounce and keeps waiting.
3. When there are no more file change events, BS then reloads the browser.

This ensures that BS will wait properly till all files have been changed before issuing the reload, and also removes the need for using the delay setting.

**Comment by shakyShane** [comment]
This is better supported in an upcoming major release. for now though, if you're happy to write a little bit of JS, you can get extremely reliable debouncing on file-watching by doing the following (this is the gist of how v3 will work)


```js
const Rx = require('rx');
const chok = require('chokidar');
const bs = require('browser-sync').create();

/**
 * Start Browsersync as normal, without files option
 */
bs.init({
    server: 'app'
})

/**
 * Create a stream of file-change events
 */
Rx.Observable.create(function(observer) {
    const watcher = chok
        .watch(['*.md', 'app/css', 'app/js'], {ignoreInitial: true})
        .on('all', function (event, file) {
        	observer.onNext({event, file});
        });
    return function () {
    	watcher.close();
    }
})
    /**
     * Wait for 2 seconds of event silence before emitting
     */
    .debounce(2000)
    /**
     * Only look at add/change events
     */
    .filter(function (x) {
    	return x.event === 'add' || x.event === 'change';
    })
    /**
     * Finally reload the browser
     */
    .subscribe(function (x) {
    	bs.reload();
    });
```

**Comment by adambuczynski** [comment]
Thanks for the snippet, I'll use that instead of the files options.

**Comment by adambuczynski** [comment]
Works like a charm!

**Comment by shakyShane** [comment]
@adambuczynski that's good as I never even tested it, all from memory :)

see how i'm doing it in v3 here if you're interested. https://github.com/BrowserSync/browser-sync-core/blob/master/lib/plugins/watch.js

once chokidar is wrapped as an Observable, things such as this become trivial

**Comment by adambuczynski** [comment]
:+1:

**Comment by adambuczynski** [comment]
@shakyShane one thing this implementation doesn't do though is inject modified CSS files straight into the browser. Is there an API for doing that, like `bs.reload()`?

**Comment by shakyShane** [comment]
```js
/**
     * Finally reload the browser
     */
    .subscribe(function (x) {
        bs.reload(x.file);
    });
```

**Comment by adambuczynski** [comment]
So I'd check if x.file is a css file, and then pass it to `bs.reload` as a parameter, and it will refresh only the css? Will give it a go, thanks.

**Comment by adambuczynski** [comment]
Yep, works with

```js
.subscribe(x => {
    if (x.file.match(/\.css$/)) {
      sync.reload(x.file);
    }
    else {
      sync.reload();
    }
  });
```

**Comment by shakyShane** [comment]
@adambuczynski this is implemented in 2.13.0, the following will wait for 2000ms of event silence before emitting the last event that occured

```
var bs = require("./").create();

// Start the server
bs.init({
    files: ["app"],
    reloadDebounce: 2000
});
```

**Comment by adambuczynski** [comment]
Ah great, thank you!
